### PR TITLE
feat(x/data): add resolver API

### PR DESCRIPTION
**Pull request by aaronc**
## Description

Closes: #572 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/data/v1alpha2/state.proto [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  // id is the compact data ID.
+  bytes data_id = 1;
+
+  // signer is the account address of the signer.
+  string signer = 2;
+
+  // timestamp is the time at which the signer signed this data object.
+  google.protobuf.Timestamp timestamp = 3;
+}
```
Comment (ryanchristo) [inline proto/regen/data/v1alpha2/state.proto:55]:
```suggestion
// DataSigner is a join table for associating data IDs and signers.
message DataSigner {
  option (cosmos.orm.v1alpha1.table) = {
    id: 3
    primary_key: {fields: "id,signer"}
  };

  // id is the compact data ID.
  bytes id = 1;

  // signer is the account address of the signer.
  string signer = 2;

  // timestamp is the time at which the signer signed this data object.
  google.protobuf.Timestamp timestamp = 3;
}
```

File: proto/regen/data/v1alpha2/state.proto [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    id: 5
+    primary_key: {fields: "data_id,resolver_id"}
+  };
+
+  // data_id is the compact data ID.
+  bytes data_id = 1;
+
+  // resolver_id is the ID of the resolver.
+  uint64 resolver_id = 2;
+}
```
Comment (ryanchristo) [inline proto/regen/data/v1alpha2/state.proto:90]: Either way but I think `id` is ok here as well given the message is prefixed with `Data`.

```suggestion
// DataResolver is a join table between data objects and resolvers and indicates
// that a resolver claims to be able to resolve this data object.
message DataResolver {
  option (cosmos.orm.v1alpha1.table) = {
    id: 5
    primary_key: {fields: "id,resolver_id"}
  };

  // id is the compact data ID.
  bytes id = 1;

  // resolver_id is the ID of the resolver.
  uint64 resolver_id = 2;
}
```

File: x/data/server/msg_server.go [hunk 3]
```diff
@@ -123,3 +126,54 @@ func (s serverImpl) SignData(goCtx context.Context, request *data.MsgSignData) (
...
+	for _, datum := range msg.Data {
+		_, id, _, err := s.anchorAndGetIRI(sdk.UnwrapSDKContext(ctx), datum)
+		if err != nil {
+			return nil, err
+		}
+		err = s.stateStore.DataResolverStore().Save(
+			ctx,
+			&datav1alpha2.DataResolver{
+				ResolverId: msg.ResolverId,
+				DataId:     id,
```
Comment (ryanchristo) [inline x/data/server/msg_server.go:171]: If you choose to change above..

```suggestion
				Id:     id,
```

**Review by ryanchristo — APPROVED** [review]
Looks great. Should we open an issue to add tests or are you planning to do so here?

**Comment by aaronc** [comment]
> Looks great. Should we open an issue to add tests or are you planning to do so here?

I'm hoping to add tests here... but haven't had time

File: x/data/server/testsuite/suite.go [hunk 4]
```diff
@@ -210,3 +216,41 @@ func (s *IntegrationTestSuite) TestRawDataScenario() {
 	require.NotNil(queryRes)
 	require.Equal(ts, queryRes.Entry.Timestamp)
 }
+
+func (s *IntegrationTestSuite) TestResolver() {
```
Comment (technicallyty) [inline x/data/server/testsuite/suite.go:220]: should we be checking more here? i.e. checking query responses are equal to msg server requests

**Review by technicallyty — COMMENTED** [review]
good work! just a few comments

File: x/data/server/server.go [hunk 5]
```diff
@@ -1,16 +1,28 @@
...
 	servermodule "github.com/regen-network/regen-ledger/types/module/server"
 	"github.com/regen-network/regen-ledger/x/data"
 	"github.com/regen-network/regen-ledger/x/data/server/lookup"
 )
 
+var ModuleSchema = ormdb.ModuleSchema{
+	FileDescriptors: map[uint32]protoreflect.FileDescriptor{
+		1: datav1alpha2.File_regen_data_v1alpha2_state_proto,
+	},
+	Prefix: []byte{ORMStatePrefix},
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/server.go:19]: We need a prefix to not collide with existing non-ORM keys

File: x/data/server/query_server.go [hunk 6]
```diff
@@ -122,3 +121,57 @@ func (s serverImpl) Signers(goCtx context.Context, request *data.QuerySignersReq
...
+
+func (s serverImpl) Resolvers(ctx context.Context, request *data.QueryResolversRequest) (*data.QueryResolversResponse, error) {
+	id, err := s.getID(ctx, request.Iri)
+	if err != nil {
+		return nil, err
+	}
+
+	it, err := s.stateStore.DataResolverStore().
+		List(ctx, datav1alpha2.DataResolverPrimaryKey{}.WithId(id),
+			ormlist.Paginate(ormstore.GogoPageReqToPulsarPageReq(request.Pagination)))
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply x/data/server/query_server.go:136]: https://github.com/regen-network/regen-ledger/pull/708/commits/3748fc1d6b66cf31dc3f0128f1c84a34baff1430

File: x/data/server/query_server.go [hunk 7]
```diff
@@ -122,3 +121,57 @@ func (s serverImpl) Signers(goCtx context.Context, request *data.QuerySignersReq
...
+	if err != nil {
+		return nil, err
+	}
+
+	res := &data.QueryResolversResponse{}
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply x/data/server/query_server.go:141]: https://github.com/regen-network/regen-ledger/pull/708/commits/c98c52d6cfd7e7bc82866bf764f2f1b314f88111

File: x/data/server/testsuite/suite.go (hunk 4 above)
Comment (ryanchristo) [reply x/data/server/testsuite/suite.go:219]: https://github.com/regen-network/regen-ledger/pull/708/commits/a7fc3fc42d340d349618539d8f407102ab19a470

**Comment by aaronc** [comment]
How close is this @ryanchristo ?

**Comment by ryanchristo** [comment]
> How close is this @ryanchristo ?

Trying to sort out the failed CI checks. Aside from that, this is ready for another review. cc @technicallyty

File: proto/regen/data/v1alpha2/state.proto [hunk 8]
```diff
@@ -0,0 +1,90 @@
+syntax = "proto3";
+
+package regen.data.v1alpha2;
```
Comment (technicallyty) [inline proto/regen/data/v1alpha2/state.proto:3]: this isn't being used yet is it? should it go in v1alpha2? or would this be for a v1 release?

**Review by technicallyty — APPROVED** [review]
lgtm, one question

File: proto/regen/data/v1alpha2/state.proto (hunk 8 above)
Comment (aaronc) [reply proto/regen/data/v1alpha2/state.proto:3]: v1 for release

Comment (ryanchristo) [reply proto/regen/data/v1alpha2/state.proto:3]: Opened #849 to handle as a follow-up

**Comment by ryanchristo** [comment]
Any ideas what might be causing the following error? cc @technicallyty @aaronc 

```
regen/ecocredit/v1beta1/baskets/state.cosmos_orm.go:95: undefined: ormlist.Prefix
regen/ecocredit/v1beta1/baskets/state.cosmos_orm.go:96: this.table.GetIndexByID(prefixKey.id()).Iterator undefined (type ormtable.Index has no field or method Iterator)
regen/ecocredit/v1beta1/baskets/state.cosmos_orm.go:101: undefined: ormlist.Start
regen/ecocredit/v1beta1/baskets/state.cosmos_orm.go:101: undefined: ormlist.End
regen/ecocredit/v1beta1/baskets/state.cosmos_orm.go:102: this.table.GetIndexByID(from.id()).Iterator undefined (type ormtable.Index has no field or method Iterator)
```

**Comment by aaronc** [comment]
Try updating the orm version and running codegen again

**Comment by ryanchristo** [comment]
> Try updating the orm version, upgrading protoc-gen-go-cosmos-orm, and running codegen again

Hmmm... no luck. Same output and a couple new errors in the group module.

**Comment by ryanchristo** [comment]
We had generated api code that outdated and not being overwritten because it was generated when we had nested basket and marketplace in the v1beta1 proto directory.

**Comment by technicallyty** [comment]
> We had generated api code that outdated and not being overwritten because it was generated when we had nested basket and marketplace in the v1beta1 proto directory.

maybe our gen script needs to wipe the API folder before it begins

**Pull request merged**